File: linked-lists/dublyLinkedList.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
    this.prev = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  push(data) {
    let newNode = new Node(data);
    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
    this.length += 1;
    return this;
  }

  pop() {
    if (this.head === null) {
      return null;
    }
    let currentTail = this.tail;
    let secondToLastNode = currentTail.prev;
    this.removeNodeBindings(currentTail);
    this.tail = secondToLastNode;
    return currentTail;
  }

  shift() {
    if (this.head === null) {
      return null;
    }
    let currentHead = this.head;
    let secondToHeadNode = currentHead.next;
    this.removeNodeBindings(currentHead);
    this.head = secondToHeadNode;
    return currentHead;
  }

  unshift(data) {
    let newNode = new Node(data);
    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.head.prev = newNode;
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length += 1;
    return this;
  }

  get(index) {
    if (index == null || index < 0 || index >= this.length) {
      return null;
    }
    let current;
    if (index < this.length / 2) {
      current = this.head;
      for (let i = 0; i !== index; i++) {
        current = current.next;
      }
    } else {
      current = this.tail;
      for (let i = this.length - 1; i !== index; i--) {
        current = current.prev;
      }
    }
    return current;
  }

  set(index, value) {
    let foundNode = this.get(index);
    if (foundNode != null) {
      foundNode.value = value;
      return true;
    }
    return false;
  }

  insert(index, value) {
    if (index < 0 || index > this.length) return false;
    if (index === 0) return !!this.unshift(value);
    if (index === this.length) return !!this.push(value);

    let newNode = new Node(value);
    let beforeNode = this.get(index - 1);
    let afterNode = beforeNode.next;

    beforeNode.next = newNode, newNode.prev = beforeNode;
    newNode.next = afterNode, afterNode.prev = newNode;
    this.length++;
    return true
  }

  remove(index) {
    if (index < 0 || index >= this.length) return undefined;
    if (index === 0) return this.shift();
    if (index === this.length - 1) return this.pop();
    
    let removedNode = this.get(index);
    let beforeNode = removedNode.prev;
    let afterNode = removedNode.next;
    beforeNode.next = afterNode;
    afterNode.prev = beforeNode;
    this.removeNodeBindings(removedNode);
  }

  removeNodeBindings(currentNode) {
    if (currentNode.prev !== null) {
      currentNode.prev.next = currentNode.next;
    }
    if (currentNode.next !== null) {
      currentNode.next.prev = currentNode.prev;
    }
    currentNode.prev = null;
    currentNode.next = null;
    this.length -= 1;
    if (this.length === 0) {
      this.head = null;
      this.tail = null;
    }
    return currentNode;
  }
}

let list = new LinkedList();

list.push(901);
list.push(902);
list.push(903);
list.push(904);
list.push(905);
list.remove(2)

File: data_structures/dubly_linked_list/dubly_linked_list.py
class Node():
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None

    def __str__(self):
        return str(self.__dict__)


class LinkedList():
    def __init__(self):
        self.head = None
        self.tail = None

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.set_head(new_node)
            return
        self.insert_after(self.tail, new_node)

    def get(self, index):
        node = self.head
        current_index = 0
        while node is not None and current_index != index:
            node = node.next
            current_index += 1
        return node

    def contains(self, value):
        node = self.head
        while node is not None and node.value != value:
            node = node.next
        return node is not None

    def set_head(self, node):
        if self.head is None:
            self.head = node
            self.tail = node
            return
        self.insert_before(self.head, node)

    def set_tail(self, node):
        if self.tail is None:
            self.set_head(node)
            return
        self.insert_after(self.tail, node)

    def insert_before(self, node, new_node):
        if new_node == self.head and new_node == self.tail:
            return
        new_node.prev = node.prev
        new_node.next = node
        if node.prev is None:
            self.head = new_node
        else:
            node.prev.next = new_node
        node.prev = new_node

    def insert_after(self, node, new_node):
        if new_node == self.head and new_node == self.tail:
            return
        new_node.prev = node
        new_node.next = node.next
        if node.next is None:
            self.tail = new_node
        else:
            node.next.prev = new_node
        node.next = new_node

    def insert_at_index(self, index, new_node):
        if index == 0:
            self.set_head(new_node)
            return
        node = self.head
        current_index = 0
        while node is not None and current_index != index:
            node = node.next
            current_index += 1
        if node is not None:
            self.insert_before(node, new_node)
        else:
            self.set_tail(new_node)

    def remove_nodes_with_value(self, value):
        node = self.head
        while node is not None:
            node_to_remove = node
            node = node.next
            if node_to_remove.value == value:
                self.remove(node_to_remove)

    def remove(self, node):
        if node == self.head:
            self.head = self.head.next
        if node == self.tail:
            self.tail = self.tail.prev
        self.remove_node_bindings(node)

    def remove_node_bindings(self, node):
        if node.prev is not None:
            node.prev.next = node.next
        if node.next is not None:
            node.next.prev = node.prev
        node.prev = None
        node.next = None

    def reverse(self):
        temp = None
        node = self.head
        while node is not None:
            temp = node.prev
            node.prev = node.next
            node.next = temp
            node = node.prev
        if temp is not None:
            self.head, self.tail = self.tail, self.head


linkedList = LinkedList()

linkedList.append(1)
linkedList.append(2)
linkedList.append(3)

print(f"head: {linkedList.head}")
print(f"tail: {linkedList.tail}")
print(linkedList.contains(3))

print("reverse linked list:")
linkedList.reverse()
print(linkedList.get(0))
print(linkedList.get(1))
print(linkedList.get(2))
